### PR TITLE
Tether assigns IP to container interface

### LINKED ISSUE #455: tether support for multiple NICs and endpoints
The tether currently has code to "Apply" an endpoint (metadata/network_interface.go currently) but no support for actually processing endpoints specified in it's config. The configuration needs to be supplied (#402) and parsed.

Acceptance:

* tether assigns ip address in network specified on the docker create command line to the correct interface in the container
* tether assigns ip address specified on the docker create command line to the correct interface in the container
* stretch: link and mac address for the new endpoint in the container are returned at the port layer. There may be no way to do this at the moment, since the tether only operates on container start, but we should at least investigate if this is possible

### PR DESCRIPTION
**Pull request by andrewtchin**
For each network endpoint, find the desired interface by using supplied PCI slot number. Set the configured IP address on the interface.

Fixes #455

### CODE REVIEW

File: cmd/tether/utils_linux.go [hunk 1]
```diff
@@ -90,6 +91,23 @@ func (t *osopsLinux) SetHostname(hostname string) error {
...
 
+func pciToMAC(pciAddr string) (string, error) {
+	addrPath := "/sys/bus/pci/devices/" + pciAddr + "/net/*/address"
+	f, err := os.Open(addrPath)
+	if err != nil {
+		return "", fmt.Errorf("unable to open addrPath %s: %s", addrPath, err)
+	}
+	defer f.Close()
+
+	buf := make([]byte, 18)
```
Comment (caglar10ur) [inline cmd/tether/utils_linux.go]: Where is that 18 value coming from?I don't think we need precise buffering here, you can just use ioutil.ReadAll here.

File: cmd/tether/utils_linux.go [hunk 2]
```diff
@@ -90,6 +91,23 @@ func (t *osopsLinux) SetHostname(hostname string) error {
 	return nil
 }
 
+func pciToMAC(pciAddr string) (string, error) {
+	addrPath := "/sys/bus/pci/devices/" + pciAddr + "/net/*/address"
```
Comment (caglar10ur) [inline cmd/tether/utils_linux.go]: path.Join( "/sys/bus/pci/devices/" , pciAddr , "/net/*/address")

File: cmd/tether/tether.go [hunk 3]
```diff
@@ -134,6 +134,14 @@ func run(loader metadata.ConfigLoader) error {
 			return errors.New(detail)
 		}
 
+		for _, v := range config.Networks {
+			if err := ops.Apply(&v); err != nil {
+				detail := fmt.Sprintf("failed to apply network endpoint config: %s", err)
+				log.Error(detail)
```
Comment (caglar10ur) [inline cmd/tether/tether.go]: Why we are keep going if we log this as an error?

File: cmd/tether/utils_linux.go (hunk 2 above)
Comment (fdawg4l) [inline cmd/tether/utils_linux.go]: The `*` makes me nervous.  Is it possible more than device will be listed?   Can you guard against it?

File: cmd/tether/tether.go (hunk 3 above)
Comment (andrewtchin) [inline cmd/tether/tether.go]: fixed

File: cmd/tether/utils_linux.go (hunk 1 above)
Comment (andrewtchin) [inline cmd/tether/utils_linux.go]: cool fixed

File: cmd/tether/utils_linux.go (hunk 2 above)
Comment (andrewtchin) [inline cmd/tether/utils_linux.go]: fixed both

Comment (hickeng) [inline cmd/tether/utils_linux.go]: We will be aliasing interfaces but a basic test with `ip link add <cidr> brd + dev eth0 label eth0:0` shows that an IP alias doesn't show here.
@andrewtchin I would suggest using ioutil.ReadDir to get an actual path name as I'm not sure how Go would handle the * in the case of multiple entries - I'm surprised it works with one honestly but I learn something new most days :)

**Comment by hickeng** [comment]
LGTM - I checked the IP aliasing impact `ip addr add ${cidr}  brd + dev eth0 label myalias` and it doesn't show up as an extra interface name so no magic to do there for supporting ip aliasing. Don't know what'll happen with macvlan.

**Comment by caglar10ur** [comment]
LGTM :sheep: :it:

**Comment by andrewtchin** [comment]
@hickeng 
`ip link add link eth0 address 00:11:11:11:11:11 eth0.1 type macvlan` was ok :-)
that appears under `/net/eth0/upper_eth0.1/address`, at least in debian

**Comment by andrewtchin** [comment]
CI isn't connected to this PR anymore for some reason, closing and reopening.

File: cmd/tether/utils_linux.go (hunk 2 above)
Comment (andrewtchin) [inline cmd/tether/utils_linux.go]: @hickeng You're right I found that * didn't work

**Pull request closed without merging**